airbyte-cdk/python/airbyte_cdk/sources/streams/checkpoint/checkpoint_reader.py

```diff
@@ -59,18 +59,18 @@ def __init__(self, stream_state: Mapping[str, Any], stream_slices: Iterable[Opti
         self._has_slices = False
 
     def next(self) -> Optional[Mapping[str, Any]]:
-        try:
-            next_slice = next(self._stream_slices)
-            self._has_slices = True
-            return next_slice
-        except StopIteration:
-            # This is used to avoid sending a duplicate state message at the end of a sync since the stream has already
-            # emitted state at the end of each slice. If we want to avoid this extra complexity, we can also just accept
-            # that every sync emits a final duplicate state
-            if self._has_slices:
-                self._state = None
+        if self._has_slices:
+            next_slice = next(self._stream_slices, None)
+            if next_slice is not None:
+                return next_slice
```

Comment:
```suggestion
```suggestion
            # Set the state to None to avoid sending a duplicate state message at the end of a sync since the stream has already emitted the state at the end of each slice. 
            # If we want to avoid this extra complexity, we can also just accept
            # that every sync emits a final duplicate state```
                return next_slice
```